microbench/src/main/java/io/netty/microbench/buffer/PooledByteBufAllocatorAlignBenchmark.java

```diff
@@ -0,0 +1,144 @@
+ * under the License.
+ */
+package io.netty.microbench.buffer;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.microbench.util.AbstractMicrobenchmark;
+
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+@State(Scope.Thread)
+@Warmup(iterations = 5, time = 100, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 5, time = 100, timeUnit = TimeUnit.MILLISECONDS)
+@Fork(5)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+public class PooledByteBufAllocatorAlignBenchmark extends
+        AbstractMicrobenchmark {
+
+    private static final Random rand = new Random();
+
+    private static final int CACHE_LINE_MAX = 256;
+
+    private static final int OFFSET_ADD = CACHE_LINE_MAX * 1337;
+
+    @Param({ "0", "64" })
+    private int cacheAlign;
+
+    @Param({ "01024", "04096", "16384", "65536", "1048576" })
+    private int size;
+
+    private int block = 4;
+
+    private PooledByteBufAllocator pooledAllocator;
+
+    private ByteBuf pooledDirectBuffer;
+
+    private byte[] bytes;
+
+    private int sizeMask;
+
+    private int alignOffset;
+
+    @Setup
+    public void doSetup() {
+        pooledAllocator = new PooledByteBufAllocator(true, 4, 4, 8192, 11, 0,
+                0, 0, true, cacheAlign);
+        pooledDirectBuffer = pooledAllocator.directBuffer(size + CACHE_LINE_MAX);
+        sizeMask = size - 1;
+        if (cacheAlign == 0) {
+            long addr = pooledDirectBuffer.memoryAddress();
+            if (addr % 64 == 0) {
+                alignOffset = 63;
+            }
+            int off = 0;
+            for (int c = 0; c < size; c++) {
+                off = (off + OFFSET_ADD) & sizeMask;
+                if ((addr + off + alignOffset) % block == 0) {
+                    throw new IllegalStateException(
+                            "Misaligned address is not really aligned");
+                }
+            }
+        } else {
+            alignOffset = 0;
+            int off = 0;
+            long addr = pooledDirectBuffer.memoryAddress();
+            for (int c = 0; c < size; c++) {
+                off = (off + OFFSET_ADD) & sizeMask;
+                if ((addr + off + alignOffset) % block != 0) {
+                    throw new IllegalStateException(
+                            "Aligned address is not really aligned");
+                }
+            }
+        }
```

Comment:
What do you mean? I use alignOffset for the miss-align line. Yes, it will be used in some cases. I didn't find the way to change offset inside buffer once it was created.